src/compute-client/src/controller.rs

```diff
@@ -163,9 +172,365 @@ impl From<anyhow::Error> for ComputeError {
+    },
+}
+
+impl ConcreteComputeInstanceReplicaLocation {
+    pub fn get_az(&self) -> Option<&str> {
+        match self {
+            ConcreteComputeInstanceReplicaLocation::Remote { .. } => None,
+            ConcreteComputeInstanceReplicaLocation::Managed {
+                availability_zone, ..
+            } => Some(availability_zone),
+        }
+    }
+}
+
+/// Resource allocations for a replica of a compute instance.
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
+pub struct ComputeInstanceReplicaAllocation {
+    /// The memory limit for each process in the replica.
+    pub memory_limit: Option<MemoryLimit>,
+    /// The CPU limit for each process in the replica.
+    pub cpu_limit: Option<CpuLimit>,
+    /// The number of processes in the replica.
+    pub scale: NonZeroUsize,
+    /// The number of worker threads in the replica.
+    pub workers: NonZeroUsize,
+}
+
+impl ComputeInstanceReplicaAllocation {
+    pub fn workers(&self) -> NonZeroUsize {
+        self.workers
+    }
+}
+
+/// Logging configuration of a replica.
+/// Changing this type requires a catalog storage migration!
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub enum ConcreteComputeInstanceReplicaLogging {
+    /// Instantiate default logging configuration upon system start.
+    /// To configure a replica without logging, ConcreteViews(vec![],vec![]) should be used.
+    Default,
+    /// Logging sources and views have been built for this replica.
+    ConcreteViews(Vec<(LogVariant, GlobalId)>, Vec<(LogView, GlobalId)>),
+}
+
+impl ConcreteComputeInstanceReplicaLogging {
+    /// Return all persisted introspection sources contained
+    pub fn get_sources(&self) -> Vec<(LogVariant, GlobalId)> {
+        match self {
+            ConcreteComputeInstanceReplicaLogging::Default => vec![],
+            ConcreteComputeInstanceReplicaLogging::ConcreteViews(logs, _) => logs.clone(),
+        }
+    }
+
+    /// Return all persisted introspection views contained
+    pub fn get_views(&self) -> Vec<(LogView, GlobalId)> {
+        match self {
+            ConcreteComputeInstanceReplicaLogging::Default => vec![],
+            ConcreteComputeInstanceReplicaLogging::ConcreteViews(_, views) => views.clone(),
+        }
+    }
+
+    /// Return all ids of the persisted introspection views contained
+    pub fn get_view_ids(&self) -> Vec<GlobalId> {
+        self.get_views().into_iter().map(|(_, id)| id).collect()
+    }
+
+    /// Return all ids of the persisted introspection sources contained
+    pub fn get_source_ids(&self) -> Vec<GlobalId> {
+        self.get_sources().into_iter().map(|(_, id)| id).collect()
+    }
+
+    /// Return all ids of the persisted introspection sources and logs contained
+    pub fn get_source_and_view_ids(&self) -> Vec<GlobalId> {
+        self.get_source_ids()
+            .into_iter()
+            .chain(self.get_view_ids().into_iter())
+            .collect()
+    }
+}
```

Comment:
Note to self: change this implementation to allocate less. I think all functions can in theory get away without allocating at all.

Reply:
https://github.com/MaterializeInc/materialize/pull/14783